File: python/tree/binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderTraversal(self, root):
        arr = []
        st = []
        cur = root
        while cur or st:
            while cur:
                st.append(cur)
                cur = cur.left
            arr.append(st[-1].val)
            cur = st.pop().right

        return arr

File: c/tree/binary-tree-inorder-traversal.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* inorderTraversal(struct TreeNode* root, int* returnSize) {
    int *arr;
    struct TreeNode **st;
    struct TreeNode *cur;
    int top = -1;
    int arr_count  = 32;
    int st_count = 32;
    int idx = 0;

    st = malloc(sizeof(struct TreeNode *) * st_count);
    arr = malloc(sizeof(int) * arr_count);

    cur = root;
    while(cur || top > -1){
        while(cur){
            if(top + 1 == st_count){
                st_count <<= 1;
                st = realloc(st,sizeof(struct TreeNode *) * st_count);
            }
            st[++top] = cur;
            cur = cur->left;
        }

        struct TreeNode *tmp = st[top--];
        cur = tmp->right;
        if(idx == arr_count){
            arr_count <<= 1;
            arr = realloc(arr, sizeof(int) * arr_count);
        }
        arr[idx++] = tmp->val;
    }

    *returnSize = idx ;
    free(st);
    return arr;
}
